### PR TITLE
Changes for swift 2.3 and 3.0

### PR DESCRIPTION
**Pull request by snit-ram**
* Compatible with Swift 3.0, 2.3 and 2.2
* Compatible with xcode 7.3 and 8 (tested using beta 6)

I'm not sure how to make travis to check against all swift/xcode versions, so I kept it running against xcode-7.3/swift-2.2 and tested the other cases locally

Great project, btw. I love Signals 😄

### CODE REVIEW

**Comment by snit-ram** [comment]
beta6 is out and there are changes to swift syntax. I'll try to do the changes

**Comment by snit-ram** [comment]
If you believe it's good to merge it may be a good idea to move forward. It works up until beta5. I can send a new one that works in beta6, but this is a good first step

**Comment by artman** [comment]
Thanks Rafel. I think you should update this PR for beta6 compatibility as everyone will be upgrading to the latest beat quickly.

**Comment by masters3d** [comment]
I moved a 10k loc swift project to use `#if swift(>=3.0)` so to be compatible with 2.2/2.3 but it became very painful to maintain. I ended up ripping out all the code and just maintaining two separate branches (2.2 and 3.0).

**Comment by chetanankola** [comment]
Seems like we are waiting for beta 6 changes for this to be merged.. but this already makes it swift 2.3 ready .. I guess the world can wait till Swift 3 beta X changes more API's but merging this will make this world a better place..

File: Signals/Signal.swift [hunk 1]
```diff
@@ -69,37 +69,67 @@ final public class Signal<T> {
     /// - parameter listener: The listener object. Sould the listener be deallocated, its associated callback is 
     ///   automatically removed.
     /// - parameter callback: The closure to invoke whenever the signal fires.
+    #if swift(>=3.0)
+    @discardableResult
+    public func listen(_ listener: AnyObject, callback: (T) -> Void) -> SignalListener<T> {
```
Comment (artman) [inline Signals/Signal.swift]: Shouldn't strive for API compatibility, but make the API Swifty. Can we rename this to
`listen(with: AnyObject, callback: ...`

File: Signals/Signal.swift [hunk 2]
```diff
@@ -69,37 +69,67 @@ final public class Signal<T> {
...
+    #endif
     
     /// Attaches a listener to the signal that is removed after the signal has fired once.
     ///
     /// - parameter listener: The listener object. Sould the listener be deallocated, its associated callback is 
     ///   automatically removed.
     /// - parameter callback: The closure to invoke when the signal fires for the first time.
+    #if swift(>=3.0)
+    @discardableResult
+    public func listenOnce(_ listener: AnyObject, callback: (T) -> Void) -> SignalListener<T> {
```
Comment (artman) [inline Signals/Signal.swift]: Shouldn't strive for API compatibility, but make the API Swifty. Can we rename this to
`listenOnce(with: AnyObject, callback: ...`

File: Signals/Signal.swift [hunk 3]
```diff
@@ -69,37 +69,67 @@ final public class Signal<T> {
...
     
     /// Attaches a listener to the signal and invokes the callback immediately with the last data fired by the signal
     /// if it has fired at least once and if the `retainLastData` property has been set to true.
     ///
     /// - parameter listener: The listener object. Sould the listener be deallocated, its associated callback is 
     ///   automatically removed.
     /// - parameter callback: The closure to invoke whenever the signal fires.
+    #if swift(>=3.0)
+    @discardableResult
+    public func listenPast(_ listener: AnyObject, callback: (T) -> Void) -> SignalListener<T> {
```
Comment (artman) [inline Signals/Signal.swift]: I've never liked this naming of this method. Would make much more to use default values to indicate whether we want to send past events now that the API is going to break anyways:

`listen(with: AnyObject, firePastEvent: Bool = false, callback: ...`

So this wouldn't need its own function, and listenOnce could also implement the firePastEvent boolean.

File: Signals/Signal.swift [hunk 4]
```diff
@@ -108,6 +138,19 @@ final public class Signal<T> {
     /// - parameter listener: The listener object. Sould the listener be deallocated, its associated callback is 
     ///   automatically removed.
     /// - parameter callback: The closure to invoke whenever the signal fires.
+    #if swift(>=3.0)
+    @discardableResult
+    public func listenPastOnce(_ listener: AnyObject, callback: (T) -> Void) -> SignalListener<T> {
```
Comment (artman) [inline Signals/Signal.swift]: This function becomes unnecessary.

File: Signals/Signal.swift [hunk 5]
```diff
@@ -118,11 +161,21 @@ final public class Signal<T> {
         }
         return signalListener
     }
+    #endif
 
+    #if swift(>=3.0)
+    public func fire(_ data: T) {
```
Comment (artman) [inline Signals/Signal.swift]: I'd remove the _ so that you have to explicitly call out the data parameter.

File: Signals/Signal.swift [hunk 6]
```diff
@@ -118,11 +161,21 @@ final public class Signal<T> {
...
+    #else
+    public func fire(data: T) {
+        _fire(data: data)
+    }
+    #endif
     /// Fires the singal.
     ///
     /// - parameter data: The data to fire the signal with.
-    public func fire(data: T) {
+    private func _fire(data data: T) {
```
Comment (artman) [inline Signals/Signal.swift]: No underscores for private functions please.

File: Signals/Signal.swift [hunk 7]
```diff
@@ -134,10 +187,19 @@ final public class Signal<T> {
         }
     }
     
+    #if swift(>=3.0)
+    public func removeListener(_ listener: AnyObject) {
```
Comment (artman) [inline Signals/Signal.swift]: Swiftify this to:

`remove(listener: AnyObject)`

File: Signals/Signal.swift [hunk 8]
```diff
@@ -134,10 +187,19 @@ final public class Signal<T> {
...
+    #else
+    public func removeListener(listener: AnyObject) {
+        _removeListener(listener: listener)
+    }
+    #endif
     /// Removes an object as a listener of the Signal.
     ///
     /// - parameter listener: The listener to remove.
-    public func removeListener(listener: AnyObject) {
+    private func _removeListener(listener listener: AnyObject) {
```
Comment (artman) [inline Signals/Signal.swift]: Won't have no underscores in private functions ;)

File: Signals/Signal.swift [hunk 9]
```diff
@@ -166,18 +232,37 @@ public class SignalListener<T> {
...
+    @discardableResult
+    private func dispatch(_ data: T) -> Bool {
+        return _dispatch(data: data)
+    }
+    #else
     private func dispatch(data: T) -> Bool {
+        return _dispatch(data: data)
+    }
+    #endif
+    private func _dispatch(data data: T) -> Bool {
```
Comment (artman) [inline Signals/Signal.swift]: Can we just copy the whole function body into the Swift 3 and Swift 2.3 versions? Yeah, it's repetition, but easier to clean up once 3.0 comes out and don't need no underscore'd private functions.

File: Signals/Signal.swift [hunk 10]
```diff
@@ -227,30 +328,54 @@ public class SignalListener<T> {
...
     }
+    #endif
     
     /// Tells the listener to queue up all signal fires until the elapsed time has passed and only once dispatch the 
     /// last received data. A delay of 0 will wait until the next runloop to dispatch the signal fire to the listener.
     /// - parameter delay: The number of seconds to delay dispatch
     /// - returns: Returns self so you can chain calls.
+    #if swift(>=3.0)
+    @discardableResult
+    public func queueAndDelayBy(_ delay: TimeInterval) -> SignalListener {
```
Comment (artman) [inline Signals/Signal.swift]: Swiftify to `queue(andDelayBy: TimeInterval)`

File: Signals/Signal.swift [hunk 11]
```diff
@@ -227,30 +328,54 @@ public class SignalListener<T> {
...
 
     /// Assigns a dispatch queue to the SignalListener. The queue is used for scheduling the listener calls. If not nil,
     /// the callback is fired asynchronously on the specified queue. Otherwise, the block is run synchronously on the
     /// posting thread (default behaviour).
     ///
     /// - parameter queue: A queue for performing the listener's calls.
     /// - returns: Returns self so you can chain calls.
+    #if swift(>=3.0)
+    @discardableResult
+    public func dispatchOnQueue(_ queue: DispatchQueue) -> SignalListener {
```
Comment (artman) [inline Signals/Signal.swift]: Swiftify to `dispatch(onQueue: DispatchQueue)`

**Comment by snit-ram** [comment]
@artman I agree with switfying the methods. I was just unaware that you were open to it. I'll do the changes. Do you prefer the Swift API guidelines to be applied only to swift 3, or to all versions?

File: Signals/Signal.swift (hunk 6 above)
Comment (snit-ram) [inline Signals/Signal.swift]: Agreed. I did it for making clear the public methods are just facades to this one, avoiding the naming conflicts and code duplication. I'll find better naming for it

File: Signals/Signal.swift (hunk 11 above)
Comment (snit-ram) [inline Signals/Signal.swift]: Should this be done for swift 2.2/2.3 as well?

File: Signals/Signal.swift (hunk 1 above)
Comment (snit-ram) [inline Signals/Signal.swift]: do you think `listen(on: AnyObject, callback: ...` might be a naming?

**Comment by snit-ram** [comment]
@artman, I did the changes for beta6 renaming the APIs to be more swifty and addressing your comments

**Comment by artman** [comment]
Cheers, @snit-ram. Looks good. Merged. I'll clean up when GM of Xcode 7 lands and push the pod.

**Comment by snit-ram** [comment]
Thanks for reviewing and merging the PR

**Comment by hpoul** [comment]
Maybe a stupid question, but is there a (released, tagged, branched) version of Signals which works with Swift 2.3? I think Signals 4.0.0 is only for Swift 3?

**Pull request merged**